src/audio/mod.rs

```diff
@@ -1,39 +1,53 @@
-use crate::{ring_buffer, Pipeline, Result};
+pub mod midi;
+
+use crate::{ring_buffer, AudioUnit, Pipeline, Result};
 use anyhow::anyhow;
 use cpal::{
     traits::{DeviceTrait, HostTrait, StreamTrait},
     Device, StreamConfig,
 };
 use ringbuf::RingBuffer;
 use std::time::Duration;
+use wmidi::MidiMessage;
 
-const LATENCY: f32 = 20.0;
+const LATENCY: f32 = 50.0;
 
 pub fn run(
     input_device: &Device,
     output_device: &Device,
     config: &StreamConfig,
+    midi_port_name: Option<String>,
     mut pipeline: Pipeline,
 ) -> Result<()> {
     let latency_num_frames = (LATENCY / 1_000.0) * config.sample_rate.0 as f32;
     let latency_num_samples = latency_num_frames as usize * config.channels as usize;
 
     let ring = RingBuffer::new(latency_num_samples * 2);
     let (mut producer, mut consumer) = ring.split();
-
     ring_buffer::write_empty_samples(&mut producer, latency_num_samples)?;
 
+    let midi_messages = match midi_port_name {
+        Some(midi_port_name) => Some(midi::listen_for_input(&midi_port_name)?),
+        None => None,
+    };
```

Comment:
I made this optional so that we can test things without having a keyboard plugged in.